frugalos_config/src/cancelable_future.rs

```diff
@@ -0,0 +1,180 @@
+        self.0.send(()).map_err(Error::from)
+    }
+}
+
+impl SignalReceiver {
+    /// signalを取り出そうとする。
+    /// このメソッドは受信できるsignalがない場合にはエラーを返す。
+    pub fn try_recv_signal(&self) -> Result<(), Error> {
+        self.0.try_recv().map_err(Error::from)
+    }
+}
+
+impl Future for SignalReceiver {
+    type Item = ();
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let result = self.try_recv_signal();
+
+        match result {
+            Ok(()) => Ok(Async::Ready(())),
+            Err(e) => match e.kind() {
+                ErrorKind::WouldBlock => Ok(Async::NotReady),
+                _ => Err(e),
+            },
+        }
+    }
+}
+
+/// Signalの送信口`SignalSender`と受信口`SignalReceiver`を生成する。
+pub fn make_channel() -> (SignalSender, SignalReceiver) {
+    let (sender, receiver) = mpsc::channel();
+    (SignalSender(sender), SignalReceiver(receiver))
+}
+
+/// `future: Future<Item = (), Error = ()>`からキャンセル可能なFutureを作るための構造体である。
+/// # Example
+///
+/// ```ignore
+/// let mut executor = track!(ThreadPoolExecutor::new().map_err(Error::from))?;
+/// let mut rpc_server_builder = RpcServerBuilder::new(node.addr);
+/// ...
+/// let rpc_server = rpc_server_builder.finish(executor.handle());
+/// let (cancelable_rpc_server, cancelizer) =
+///     Cancelable::new(rpc_server.map_err(move |e| panic!("Error: {}", e)));
+/// ...
+/// executor.spawn(cancelable_rpc_server);
+/// ...
+/// // `some_future`は終了直前に、`rpc_server`を終させるために、`cancelizer.send_signal()`を呼ぶ。
+/// executor.run_future(some_future).unwrap();
+/// ```
+pub struct Cancelable {
+    signal_rx: SignalReceiver,
+    inner: Box<Future<Item = (), Error = ()> + Send + 'static>,
+}
+
+impl Cancelable {
+    /// 受け取ったfutureをキャンセル可能なfutureに変換する。  
+    /// `let (cancelable_future, cancelizer) = Cancelable::new(future)`  
+    /// において、`cancelable_future`がキャンセル可能なfutureであり、  
+    /// このfutureをキャンセルしたくなった時点で `cancelizer.send_signal()` を呼び出せば、  
+    /// シグナル送信以降は `cancelable_future.poll() == Ok(Async::Ready())` を即時返却するようになる。
+    pub fn new<F>(future: F) -> (Self, SignalSender)
+    where
+        F: Future<Item = (), Error = ()> + Send + 'static,
+    {
+        let (signal_tx, signal_rx) = make_channel();
+        (
+            Cancelable {
+                inner: Box::new(future),
+                signal_rx,
+            },
+            signal_tx,
+        )
+    }
+}
+
+impl Future for Cancelable {
+    type Item = ();
+    type Error = ();
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        if let Async::Ready(_) = self.signal_rx.poll().unwrap() {
+            return Ok(Async::Ready(()));
+        } else {
+            self.inner.poll()
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Cancelable;
+    use fibers::{Executor, ThreadPoolExecutor};
+    use futures::{Async, Future, Poll};
+    use trackable::result::TestResult;
+    use {Error, ErrorKind};
+
+    struct S {
```

Comment:
I like a more self-descriptive name. Especially, a single uppercase alphabet looks like a type parameter at the first glance (for me).